typescript/index.d.ts

```diff
@@ -0,0 +1,160 @@
+    definitions: string,
+    maxCoverage: number,
+    useLinkers: boolean
+  ): string;
+
+  // coord generation
+  generate_aligned_coords(templateMol: JSMol): string;
+  generate_aligned_coords(templateMol: JSMol, useCoordGen: boolean): string;
+  generate_aligned_coords(
+    templateMol: JSMol,
+    useCoordGen: boolean,
+    allowOptionalAttachements: boolean
+  ): string;
+  generate_aligned_coords(
+    templateMol: JSMol,
+    useCoordGen: boolean,
+    allowOptionalAttachements: boolean,
+    acceptFailure: boolean
+  ): string;
+  is_valid(): boolean;
+  has_coords(): boolean;
+  get_stereo_tags(): string;
+  get_aromatic_form(): string;
+  get_kekule_form(): string;
+  set_new_coords(): boolean;
+  set_new_coords(useCoordGen: boolean): boolean;
+  get_new_coords(): string;
+  get_new_coords(useCoordGen: boolean): string;
+
+  // property methods
+  has_prop(key: string): boolean;
+  get_prop_list(): string[];
+  get_prop_list(includePrivate: boolean): string[];
+  get_prop_list(includePrivate: boolean, includeComputed: boolean): string[];
+  set_prop(key: string, val: string): boolean;
+  set_prop(key: string, val: string, computed: boolean): boolean;
+  get_prop(key: string): string;
+
+  // hydrogen methods
+  remove_hs(): string;
+  add_hs(): string;
+
+  // depiction methods
+  normalize_depiction(): number;
+  normalize_depiction(canonicalize: number): number;
+  normalize_depiction(canonicalize: number, scaleFactor: number): number;
+  straighten_depiction(): void;
+
+  // canvas methods
+  draw_to_canvas(
+    canvas: HTMLCanvasElement,
+    width: number,
+    height: number
+  ): void;
+  draw_to_canvas_with_highlights(
+    canvas: HTMLCanvasElement,
+    details: string
+  ): void;
+  draw_to_canvas_with_offset(
+    canvas: HTMLCanvasElement,
+    offsetx: number,
+    offsety: number,
+    width: number,
+    height: number
+  ): void;
+}
+
+export interface SubstructLibrary {
+  add_mol(m: JSMol): number;
+  add_smiles(smi: string): number;
+  add_trusted_smiles(smi: string): number;
+
+  get_mol(i: number): JSMol;
+
+  get_matches(q: JSMol): string;
+  get_matches(q: JSMol, maxResults: number): string;
+  get_matches(
+    q: JSMol,
+    useChirality: boolean,
+    numThreads: number,
+    maxResults: number
+  ): string;
+  count_matches(q: JSMol): number;
+  count_matches(q: JSMol, useChirality: boolean, numThreads: number): number;
+}
+
+export interface RDKitModule {
+  Mol: JSMol;
+  SubstructLibrary: SubstructLibrary;
+  get_mol(input: string, details_json: string): JSMol;
+  get_mol_from_pickle(pkl: string): JSMol;
+  get_mol_from_uint8array(pklAsUInt8Array: Uint8Array): JSMol;
+  get_mol_copy(other: JSMol): JSMol;
+  get_qmol(input: string): JSMol;
+  get_inchikey_for_inchi(inchi: string): string;
+
+  version(): string;
+  prefer_coordgen(prefer: boolean): void;
+  use_legacy_stereo_perception(value: boolean): void;
+}
+
+export type RDKitLoader = () => Promise<RDKitModule>;
```

Comment:
@adam-of-barot RDKitModule was changed to an interface to be able to type the promise of the RDKitLoader with the RDKitModule. 

RDKitLoader type was added too as you can see, and the Window object was extended so as to include the initRDKitModule loader and also the RDKitModule itself under the RDKit property.

Mandatory is the initRDKitModule property on Window. The RDKit property should be considered optional or experimental for now.

Reply:
I initally avoided extending the Window interface with the RDKit property, as even the official example pages have differently named properties for the module (RDKitModule in vanilla JS vs RDKit in ReactJS).